esphome/components/remote_base/nec_protocol.cpp

```diff
@@ -6,97 +6,231 @@ namespace remote_base {
+    // AGC Repeat header (shorter version of the initial AGC header)
+    dst->item(AGC_HIGH_US, SHORT_PAUSE_LOW_US);  // Shortened AGC header
+    dst->mark(BIT_HIGH_US);                      // Stop bit to complete the repeat code
+
+    // Add space after repeat code, except after the final repeat
+    if (repeats < data.repeats - 1) {
+      dst->space(SPACE_AGC_REPEAT_US);
+    }
+  }
 }
+
 optional<NECData> NECProtocol::decode(RemoteReceiveData src) {
   NECData data{
       .address = 0,
       .command = 0,
-      .command_repeats = 1,
+      .repeats = 0,  // Start with 0, as the first frame is counted explicitly
+      .type = NECCodeType::FRAME_WITH_REPEATS,
   };
-  if (!src.expect_item(HEADER_HIGH_US, HEADER_LOW_US))
+
+  // Validate the AGC header (start of frame or repeat code)
+  if (!src.expect_mark(AGC_HIGH_US)) {
+    return {};
+  }
+
+  // Validate the short pause (repeat code)
+  if (src.expect_space(SHORT_PAUSE_LOW_US)) {
+    // Validate the stop bit of repeat code
+    if (!src.expect_mark(BIT_HIGH_US)) {
+      return {};
+    }
+
+    // Repeat code received
+    return NEC_REPEAT_CODE_DATA;
+  }
+
+  // Validate the long pause (message frame)
+  if (!src.expect_space(LONG_PAUSE_LOW_US)) {
     return {};
+  }
 
+  // Validate address bits
   for (uint16_t mask = 1; mask; mask <<= 1) {
     if (src.expect_item(BIT_HIGH_US, BIT_ONE_LOW_US)) {
-      data.address |= mask;
+      data.address |= mask;  // Logic '1'
     } else if (src.expect_item(BIT_HIGH_US, BIT_ZERO_LOW_US)) {
-      data.address &= ~mask;
+      // Logic '0', since the address is already initialized with 0, `data.address &= ~mask;` is not needed
     } else {
       return {};
     }
   }
 
+  // Validate command bits
   for (uint16_t mask = 1; mask; mask <<= 1) {
     if (src.expect_item(BIT_HIGH_US, BIT_ONE_LOW_US)) {
-      data.command |= mask;
+      data.command |= mask;  // Logic '1'
     } else if (src.expect_item(BIT_HIGH_US, BIT_ZERO_LOW_US)) {
-      data.command &= ~mask;
+      // Logic '0', since the command is already initialized with 0, `data.command &= ~mask;` is not needed
     } else {
       return {};
     }
   }
 
-  while (src.peek_item(BIT_HIGH_US, BIT_ONE_LOW_US) || src.peek_item(BIT_HIGH_US, BIT_ZERO_LOW_US)) {
-    uint16_t command = 0;
-    for (uint16_t mask = 1; mask; mask <<= 1) {
-      if (src.expect_item(BIT_HIGH_US, BIT_ONE_LOW_US)) {
-        command |= mask;
-      } else if (src.expect_item(BIT_HIGH_US, BIT_ZERO_LOW_US)) {
-        command &= ~mask;
-      } else {
-        return {};
-      }
-    }
-
-    // Make sure the extra/repeated data matches original command
-    if (command != data.command) {
-      return {};
-    }
+  // Validate the final stop bit (end of the message frame)
+  if (!src.expect_mark(BIT_HIGH_US)) {
+    return {};
+  }
 
-    data.command_repeats += 1;
+  // Message frame received, `data.type = NECCodeType::FRAME` is already set
```

Comment:
Gotta fix some comments
```suggestion
  // Message frame received, `data.type = NECCodeType::FRAME_WITH_REPEATS` is already set
```